offixsdk/src/main/java/org/aerogear/offixsdk/ClientGenerator.kt

```diff
@@ -0,0 +1,139 @@
+*/
+
+class ClientGenerator(val serverUrl: String) {
+
+    private val TAG = javaClass.simpleName
+    private val applicationContext: Context = getContext()
+    val SQL_CACHE_NAME = "oclientDB"
+    private var apClient: ApolloClient? = null
+    private var httpClient: OkHttpClient? = null
+    private var conflictInterceptor: Interceptor? = null
+
+    fun getApolloClient(): ApolloClient? {
+
+        val apolloSqlHelper = ApolloSqlHelper(applicationContext, SQL_CACHE_NAME)
+
+        //Used Normalized Disk Cache: Per node caching of responses in SQL.
+        //Persists normalized responses on disk so that they can used after process death.
+        val cacheFactory = LruNormalizedCacheFactory(EvictionPolicy.NO_EVICTION)
+            .chain(SqlNormalizedCacheFactory(apolloSqlHelper))
+
+        //If Apollo Client is not null, return it else make a new Apollo Client.
+        //Helps in singleton pattern.
+        apClient?.let {
+            return it
+        } ?: kotlin.run {
+            apClient = ApolloClient.builder()
+                .okHttpClient(getOkhttpClient()!!)
+                .normalizedCache(cacheFactory, cacheResolver())
+                .subscriptionTransportFactory(
+                    WebSocketSubscriptionTransport.Factory(
+                        serverUrl,
+                        getOkhttpClient()!!
+                    )
+                )
+                .serverUrl(serverUrl)
+//              .defaultHttpCachePolicy(HttpCachePolicy.CACHE_FIRST)
+                .build()
+        }
+
+        return apClient
+    }
+
+    private fun getOkhttpClient(): OkHttpClient? {
+
+        //Adding HttpLoggingInterceptor() to see the response body and the results.
+        httpClient?.let {
+            return it
+        } ?: kotlin.run {
+            httpClient = OkHttpClient.Builder()
+                .addInterceptor(LoggingInterceptor())
+                .addInterceptor(getResponseInterceptor(applicationContext)!!)
+                .build()
+        }
+        return httpClient
+    }
+
+    //function to get the response after query/mutation is performed, can get conflict messages and so on.
+    private fun getResponseInterceptor(context: Context): Interceptor? {
+
+        conflictInterceptor?.let {
+            return it
+        } ?: kotlin.run {
+            conflictInterceptor = Interceptor {
+                val request = it.request()
+
+                //all the HTTP work happens, producing a response to satisfy the request.
+                val response = it.proceed(request)
+
+                //https://stackoverflow.com/a/33862068/10189663
+
+                val responseBody = response.body()
+                val bufferedSource = responseBody?.source()
+                bufferedSource?.request(Long.MAX_VALUE)
+                val buffer = bufferedSource?.buffer()
+                val responseBodyString = buffer?.clone()?.readString(Charset.forName("UTF-8")) ?: ""
```

Comment:
Do you know - Why we use clone here?

Reply:
It is closeable, so can be read only once. But we want to access it again. So I found this solution to read the body again.

Reply:
I think we will need to investigate how other interceptors get access to response body - cache one for example.